File: uniffi_bindgen/src/bindings/python/templates/Protocol.py
{# misnamed - a generic "abstract base class". Used as both a protocol and an ABC for traits. #}
class {{ protocol.name }}({{ protocol.base_classes|join(", ") }}):
    {{ protocol.docstring|docstring(4) -}}
    {%- for meth in protocol.methods.iter() %}
    {%- let callable = meth.callable %}
    {% if callable.is_async %}async {% endif %}def {{ meth.callable.name }}(self, {% include "CallableArgs.py" %}) -> {{ callable.return_type.type_name }}:
        {{ meth.docstring|docstring(8) -}}
        raise NotImplementedError
    {%- else %}
    pass
    {%- endfor %}

File: uniffi_bindgen/src/bindings/swift/templates/Protocol.swift
{%- call swift::docstring_value(protocol_docstring, 0) %}
{%- if protocol_class_only %}
public protocol {{ protocol_name }}: AnyObject, Sendable {
{%- else %}
public protocol {{ protocol_name }}: Sendable {
{%- endif %}
    {% for meth in methods.iter() -%}
    {%- call swift::docstring(meth, 4) %}
    func {{ meth.name()|fn_name }}({% call swift::arg_list_protocol(meth) %}) {% call swift::is_async(meth) -%}{% call swift::throws(meth) -%}
    {%- match meth.return_type() -%}
    {%- when Some with (return_type) %} -> {{ return_type|type_name -}}
    {%- else -%}
    {%- endmatch %}
    {% endfor %}
}
